### PR TITLE
fix: Add toBuilder() to SdkHttpRequest to easily allow customizations

### PR DESCRIPTION
**Pull request by sichanyoo**
## Issue \#

https://github.com/smithy-lang/smithy-swift/issues/583

## Description of changes

Adds a convenience method `toBuilder()` to `SdkHttpRequest` to allow customers to easily obtain a builder object that has all the fields from `SdkHttpRequest`. This will allow customers to update / customize any fields in the builder, then call `build()` on the builder to obtain a mutated `SdkHttpRequest` object they want.

## Scope

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift [hunk 1]
```diff
@@ -25,6 +25,18 @@ public class SdkHttpRequest {
...
     }
+    
+    public func toBuilder() -> SdkHttpRequestBuilder {
+        return SdkHttpRequestBuilder()
+            .withBody(self.body)
+            .withMethod(self.method)
+            .withHeaders(self.headers)
+            .withPath(self.path)
+            .withHost(self.host)
+            .withQueryItems(self.queryItems ?? [])
```
Comment (jbelkins) [inline Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift:36]: Nil query items & empty query items are not the same thing.  We should preserve the `queryItems` value in the original request as-is.

**Review by jbelkins — CHANGES_REQUESTED** [review]

File: Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift (hunk 1 above)
Comment (sichanyoo) [reply Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift]: Changed to handle the nil `queryItems` value correctly.

**Review by jbelkins — APPROVED** [review]

File: Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift [hunk 2]
```diff
@@ -25,6 +25,21 @@ public class SdkHttpRequest {
...
+            .withMethod(self.method)
+            .withHeaders(self.headers)
+            .withPath(self.path)
+            .withHost(self.host)
+            .withPort(self.endpoint.port)
+            .withProtocol(self.endpoint.protocolType ?? .https)
+        if let qItems = self.queryItems {
+            builder.withQueryItems(qItems)
+        }
+        return builder
```
Comment (dayaffe) [inline Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift:41]: Does passing an empty list cause problems? We could do instead:

```
return SdkHttpRequestBuilder()
            .withBody(self.body)
            .withMethod(self.method)
            .withHeaders(self.headers)
            .withPath(self.path)
            .withHost(self.host)
            .withPort(self.endpoint.port)
            .withProtocol(self.endpoint.protocolType ?? .https)
            .withQueryItems(qItems ?? [])
```

**Review by dayaffe — CHANGES_REQUESTED** [review]
Please fix swiftlint and see if toBuilder can be simplified. If not then all good!

File: Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift (hunk 2 above)
Comment (sichanyoo) [reply Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift:41]: That's actually what I did originally, but form Josh's comment having empty query items and having nil for query items is different - so I changed it to current state.

File: Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift [hunk 3]
```diff
@@ -25,6 +25,21 @@ public class SdkHttpRequest {
         self.endpoint = endpoint
         self.body = body
     }
+
```
[the remark this answers is not in the document]
Comment (sichanyoo) [reply Sources/ClientRuntime/Networking/Http/SdkHttpRequest.swift:28]: Fixed.

**Review by dayaffe — APPROVED** [review]
LGTM

**Pull request merged**